internal/gitlab/codequality.go

```diff
@@ -0,0 +1,117 @@
+// Package gitlab generates GitLab Code Quality reports.
+package gitlab
+
+import (
+	"crypto/sha256"
+	"encoding/json"
+	"fmt"
+
+	"github.com/google/yamlfmt"
+)
+
+// CodeQuality represents a single code quality finding.
+//
+// Documentation: https://docs.gitlab.com/ee/ci/testing/code_quality.html#code-quality-report-format
+type CodeQuality struct {
+	Description string
+	Name        string
+	Fingerprint string
+	Severity    Severity
+	Path        string
+}
+
+// NewCodeQuality creates a new CodeQuality object from a yamlfmt.FileDiff.
+//
+// If the file did not change, i.e. the diff is empty, an empty struct and false is returned.
+func NewCodeQuality(diff yamlfmt.FileDiff) (CodeQuality, bool) {
+	if !diff.Diff.Changed() {
+		return CodeQuality{}, false
+	}
+
+	return CodeQuality{
+		Description: "Not formatted correctly, run yamlfmt to resolve.",
+		Name:        "yamlfmt",
+		Fingerprint: fingerprint(diff),
+		Severity:    Major,
+		Path:        diff.Path,
+	}, true
+}
+
+// Marshals a CodeQuality object into JSON.
+func (cq CodeQuality) MarshalJSON() ([]byte, error) {
+	data, err := json.Marshal(wrap(cq))
+	if err != nil {
+		return nil, fmt.Errorf("json.Marshal: %w", err)
+	}
+
+	return data, nil
+}
+
+// UnmarshalJSON unmarshals JSON into a CodeQuality object.
+func (cq *CodeQuality) UnmarshalJSON(data []byte) error {
+	var ext codeQuality
+	if err := json.Unmarshal(data, &ext); err != nil {
+		return fmt.Errorf("json.Unmarshal: %w", err)
+	}
+
+	*cq = unwrap(ext)
+
+	return nil
+}
+
+// codeQuality is the external representation of CodeQuality.
+// It is needed to add custom JSON marshaling and unmarshaling logic.
+type codeQuality struct {
```

Comment:
Is there a reason that these struct tags and the `Location` field can't just go directly on `CodeQuality`, eliminating the need for this struct? Is it mainly convenience for the way the `Location` field wraps the `Path` field?

Reply:
I lifted this code from another project that also reported line ranges. In that case the "wire format" is a bit cumbersome to use. Since we don't need this here, I have removed the custom JSON (un)marshalling completely, making the code much simpler in the process.